### ISSUE TITLE
Elastic Search 7 - Search Adapter strange bug on Magento 2.4 fresh install

### ISSUE THREAD

**Issue by cezard** [magento/magento2#29777]
### Preconditions (*)

Installed fresh copy of latest magento version from here: https://magento.com/tech-resources/download#archive-releases and included the sample data. 

PHP 7.3, MySQL 5.7.31

Link to installation: https://magento2.vali.softwareimc.dev/

### Steps to reproduce (*)

We are unable to see the Toolbar on the category page: 

https://magento2.vali.softwareimc.dev/men/tops-men.html

### Expected result (*)

The toolbar to appear on the category page : 

![image](https://user-images.githubusercontent.com/1347237/91453435-62435e80-e888-11ea-8d9e-9808177739a8.png)

### Actual result (*)

Toolbar does not appear on the category page:

![image](https://user-images.githubusercontent.com/1347237/91450335-daa82080-e884-11ea-89ce-b521eb49583f.png)

Debugging done in thread, problem found to be with the Elastic Search 7 module: https://github.com/magento/magento2/issues/29777#issuecomment-683781970

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by gwharton** [comment]
are you using Varnish? That can cause the toolbar to go missing. There is a fix merged recently that fixed the dissapearing toolbar for me with Varnish 6.

**Comment by cezard** [comment]
We are not using Varnish - it's a standard cpanel server

**Comment by rdudhara** [comment]
I have also facing the same issues.

**Comment by shikhamis11** [comment]
hi @cezard, 
can you share the details about the magento version you are using ? 
it is `2.4` or something else ?

**Comment by cezard** [comment]
It's 2.4 - fresh installation, no themes no modules

**Comment by gwharton** [comment]
I see the menu bar fine on mobile and desktop.

**Comment by cezard** [comment]
Maybe it's something server-related, but I have no idea what it could be. 

During debugging we've managed to reach the conclusion that the method getSize() on the collection returns 0, even though the count() method on the same collection returns 6 for example

**Comment by gwharton** [comment]
https://magento.stackexchange.com/questions/4036/difference-between-getsize-and-count-on-collection#:~:text=There%20is%20a%20difference.,and%20only%20override%20getSelectCountSql()%20.

**Comment by cezard** [comment]
Reindexing was the first thing I attempted - it didn't fix it.

**Comment by cezard** [comment]
Did some more debugging - it seems that the issue starts with the _totalRecords for the collection coming up as 0 from somewhere. 

The getSize on the Collection calls this method in `Magento\Framework\Data\Collection\AbstractDb`

`public function getSize()
    {
        if ($this->_totalRecords === null) {
            $sql = $this->getSelectCountSql();
            $this->_totalRecords = $this->_totalRecords ?? $this->getConnection()->fetchOne($sql, $this->_bindParams);
        }
        return (int)$this->_totalRecords;
    }`

However, since `$this->_totalRecords` is already 0, it returns it as 0. If I remove the checks for it being null and isset shorthand, the problem gets fixed, however, I'm afraid this would impact speed. 

Going deeper into finding where `_totalRecords` is set, it seems it's being set when the filters are being rendered in the `renderFiltersBefore()` method in `Magento\CatalogSearch\Model\ResourceModel\Fulltext\Collection` class:

`    protected function _renderFiltersBefore()
    {
        if ($this->isLoaded()) {
            return;
        }

        if ($this->searchRequestName !== 'quick_search_container'
            || strlen(trim($this->queryText))
        ) {
            $this->prepareSearchTermFilter();
            $this->preparePriceAggregation();

            $searchCriteria = $this->getSearchCriteriaResolver()->resolve();
            try {
                $this->searchResult =  $this->getSearch()->search($searchCriteria);
                $this->_totalRecords = $this->getTotalRecordsResolver($this->searchResult)->resolve();
            } catch (EmptyRequestDataException $e) {
                $this->searchResult = $this->createEmptyResult();
            } catch (NonExistingRequestNameException $e) {
                $this->_logger->error($e->getMessage());
                throw new LocalizedException(__('An error occurred. For details, see the error log.'));
            }
        } else {
            $this->searchResult = $this->createEmptyResult();
        }

        $this->getSearchResultApplier($this->searchResult)->apply();
        parent::_renderFiltersBefore();
    }`

Even though the `$this->_searchresult` shows a total of 24 items, the total_count is 0, so the resolver then return 0.

Any ideas on what to look for next ?

**Comment by cezard** [comment]
I managed to track it down. This is very strange though: 

Apparently the bug is in `Magento\Elasticsearch7\SearchAdapter\Adapter`

When returning the results for the search query, this is the code that returns the total : 

```
$queryResponse = $this->responseFactory->create(
            [
                'documents' => $rawDocuments,
                'aggregations' => $aggregationBuilder->build($request, $rawResponse),
                'total' => $rawResponse['hits']['total']['value'] ?? 0
            ]
        );
```

But the `$rawResponse['hits']['total']` is not an array, it contains the value directly: 

```
Array
(
    [took] => 2
    [timed_out] => 
    [_shards] => Array
        (
            [total] => 5
            [successful] => 5
            [skipped] => 0
            [failed] => 0
        )

    [hits] => Array
        (
            [total] => 25
            [max_score] => 
            [hits] => Array
                (
                    [0] => Array
                        (
                            [_index] => magento2_product_1_v49
                            [_type] => document
                            [_id] => 1874
                            [_score] => 
                            [sort] => Array
                                (
                                    [0] => 10000
                                )

                        )
                 etc
```

As soon as I change the total to:  `'total' => $rawResponse['hits']['total'] ?? 0` everything works correctly (pagination, getSize, etc).

**Comment by gwharton** [comment]
You may want to update the title so the ES team will pick it up. Strange though as I am running ES7 and I dont get the issue. I wonder if there has been a change in the ES setup between versions. Good spot though, and good debugging.

**Comment by engcom-Delta** [comment]
Hi @cezard thank you for your report, I'm not able to reproduce issue by steps you described on clean 2.4-develop
**Result:**
:heavy_check_mark: The toolbar to appear on the category page 
![image](https://user-images.githubusercontent.com/51681379/102978147-ba276480-450c-11eb-993e-2f252d10ece9.png)
![image](https://user-images.githubusercontent.com/51681379/102977802-31102d80-450c-11eb-9424-559688e7bc9c.png)

Issue is not reproducible on test instance too. Is there additional details that helps to reproduce issue?

**Comment by engcom-Delta** [comment]
@cezard we are closing this issue due to inactivity. If you'd like to update it, please reopen the issue.

**Comment by evs-xsarus** [comment]
Perhaps a change in ES ? When running this with 7.9.0, an array is returned. On ES 7.10.2, an int is returned.

**Comment by evs-xsarus** [comment]
@cezard could you confirm your versions and reopen this ticket?

**Comment by rahulks16** [comment]
It is an ElasticSearch7/Magento2 issue.

Take a look in vendor/magento/module-elasticsearch-7/SearchAdapter/Adapter.php

public function query(RequestInterface $request) : QueryResponse

$rawResponse['hits']['total']['value'] is not existent sometimes (why so ever).

The total value is found then in $rawResponse['hits']['total']

I guess I have to build a plugin (its a public function) and check the value. Something like that

```
if (isset($rawResponse['hits']['total']['value'])) {
    $total = $rawResponse['hits']['total']['value'];
}
elseif (isset($rawResponse['hits']['total'])) {
    $total = $rawResponse['hits']['total'];
}
```
After changing
```

/* modification */
		if (isset($rawResponse['hits']['total']['value'])) {
			$total = $rawResponse['hits']['total']['value'];
		}
		elseif (isset($rawResponse['hits']['total'])) {
			$total = $rawResponse['hits']['total'];
		}
		/* modification */

        $rawDocuments = $rawResponse['hits']['hits'] ?? [];
        $queryResponse = $this->responseFactory->create(
            [
                'documents' => $rawDocuments,
                'aggregations' => $aggregationBuilder->build($request, $rawResponse),
                //'total' => $rawResponse['hits']['total']['value'] ?? 0
                'total' => $total
            ]
        );
```

Now it works fine

**Comment by Hawksama** [comment]
I can confirm @rahulks16 fix works in my Magento 2.4.2 instance!

**Comment by pauldonnelly** [comment]
Thanks @rahulks16 your fix worked for me in 2.4.3-p1.
